src/lib.rs

```diff
@@ -78,3 +82,65 @@ pub fn from_value<T: serde::de::DeserializeOwned>(value: JsValue) -> Result<T> {
 pub fn to_value<T: serde::ser::Serialize + ?Sized>(value: &T) -> Result<JsValue> {
     value.serialize(&Serializer::new())
 }
+
+/// A wrapper around a [`JsValue`] (or anything that can be cast into one)
+/// that makes it pass through serialization and deserialization unchanged.
+///
+/// # Example
+/// ```rust
+/// use serde_wasm_bindgen::{PreservedValue, to_value};
+/// #[derive(serde::Serialize)]
+/// struct MyStruct {
+///     int_field: i32,
+///     js_field: PreservedValue<JsValue>,
+/// }
+/// let big_array = js_sys::Int8Array::new_with_length(1000);
+/// let s = MyStruct {
+///     int_field: 5,
+///     js_field: PreservedValue(big_array.into()),
+/// };
+///
+/// // Will return a JsValue representing an object with two fields (`int_field` and `js_field`).
+/// // `js_field` will be an `Int8Array` pointing to the same underlying JavaScript object as `big_array`.
+/// to_value(&s);
+/// ```
+#[derive(Clone, Debug, PartialEq)]
+pub struct PreservedValue<T: JsCast>(pub T);
+
+// Some arbitrary string that no one will collide with unless they try.
+pub(crate) const PRESERVED_VALUE_MAGIC: &str = "1fc430ca-5b7f-4295-92de-33cf2b145d38";
+
+#[derive(serde::Serialize, serde::Deserialize)]
+#[serde(rename = "1fc430ca-5b7f-4295-92de-33cf2b145d38")]
+struct PreservedValueWrapper(u32);
+
+impl<'de, T: JsCast> serde::Deserialize<'de> for PreservedValue<T> {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        PreservedValueWrapper::deserialize(deserializer).and_then(|wrapper| {
+            // When used with our deserializer this unsafe is correct, because the
+            // deserializer just converted a JsValue into_abi.
+            // With other deserializers, this may be incorrect but it shouldn't be UB
+            // because JsValues are represented using indices into a JS-side (i.e.
+            // bounds-checked) array.
+            let val: JsValue = unsafe { FromWasmAbi::from_abi(wrapper.0) };
+            val.dyn_into()
+                .map(|val| PreservedValue(val))
+                .map_err(|e| D::Error::custom(format!("incompatible JS value {e:?}")))
+        })
+    }
+}
+
+impl<T: JsCast> serde::Serialize for PreservedValue<T> {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let idx = self.0.as_ref().clone().into_abi();
```

Comment:
I see `.into_abi()` is implemented for `&JsValue` as well. Is the `.clone()` necessary here for ownership reasons?

If so, can you add an explanatory comment + a regression test so that someone doesn't remove it by accident?

Reply:
Yes, I will add a comment. The current tests already exercise this, in the sense that getting rid of the `clone` causes a panic about heap corruption.